Kitodo-DataFormat/src/main/java/org/kitodo/dataformat/access/OrderAwareList.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * (c) Kitodo. Key to digital objects e. V. <contact@kitodo.org>
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
+ */
+
+package org.kitodo.dataformat.access;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.function.Function;
+
+/**
+ * A list that is aware of and enforces the order of its members prescribed by
+ * the ordinal number of the media units.
+ */
+class OrderAwareList<T> extends ArrayList<T> {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Determines whether the order is ensured when Get is called. During the
+     * order check, this value is set to false to avoid an infinite loop.
+     */
+    private boolean ensureOrderOnGet = true;
+
+    /**
+     * Possibility to get an order number for T.
+     */
+    private Function<T, Integer> orderGetter;
+
+    /**
+     * Constructor for an order-aware list.
+     * 
+     * @param orderGetter
+     *            possibility to get an order number for T
+     */
+    public OrderAwareList(Function<T, Integer> orderGetter) {
+        this.orderGetter = orderGetter;
+    }
+
+    /**
+     * Sorts the list by the media unit’s order value and can handle duplicate
+     * occurrences of the same number (unlike a TreeMap). The order of list
+     * items is changed only when necessary. The order of list members with the
+     * same ordinal number is not affected.
+     * 
+     * <p>
+     * The mathematical complexity of this algorithm is linear (n) if the order
+     * is already correct, which is already given in the vast majority of
+     * invocations of the method.
+     */
+    private void ensureOrder() {
```

Comment:
Why don't you use the Comparable interface? Writing your own sorting algorithm I usually not a good idea!

Reply:
Comparable just compares and does not sort. The problem at this point is that there may be several entries that are equal in comparison. In a collection implementation, they would overwrite each other, which is not wanted. One would have to implement a separate class of a collection (HashSet, TreeSet, etc.). The Comparable is the smaller half of this problem. The algorithm is very simple, I know that from the time I programmed Basic. If you would like to change that, we would have to solve this in an own ticket, that would be a major programming task.